src/OpenFOAM/meshes/meshShapes/face/faceFunctors.H

```diff
@@ -158,4 +159,123 @@ struct faceSweptVolFunctor
     }
 };
 
+struct faceRayFunctor
+{
+    const point* meshPoints;
+    const intersection::algorithm alg;
+    const intersection::direction dir;
+
+    faceRayFunctor
+    (
+        const point* _meshPoints,
+        const intersection::algorithm _alg,
+        const intersection::direction _dir
+    ):
+        meshPoints(_meshPoints),
+        alg(_alg),
+        dir(_dir)
+    {}
+
+    __host__ __device__
+    pointHit operator()(const faceData& tgtFace, const point& p, const vector& n)
+    {
+        const label start = tgtFace.start();
+        const label nPoints = tgtFace.size();
+
+        // If the face is a triangle, do a direct calculation
+        if (nPoints == 3)
+        {
+            return triPointRef
+            (
+                meshPoints[start],
+                meshPoints[start + 1],
+                meshPoints[start + 2]
+            ).ray(p, n, alg, dir);
+        }
+
+        // Store the points belonging to the tgtFace
+        pointField facePoints(tgtFace.size());
+        // Copy the points belonging to tgtFace into facePoints
+        tgtFace.points(meshPoints, facePoints.data());
+
+        // I guess this is different from the return value of faceData::centre()
+        // Otherwise, we could just use faceData::centre() here
+        point ctr = Foam::average(facePoints);
+
+        scalar nearestHitDist = GREAT;
+
+        scalar nearestMissDist = GREAT;
+        bool eligible = false;
+
+        // Initialize to miss, distance = GREAT
+        pointHit nearest(p);
+
+        point nextPoint = ctr;
+
+        for (label pI = 0; pI < nPoints; pI++)
+        {
+            // Replace the original line : nextPoint = meshPoints[f[fcIndex(pI)]];
+            // with the following line.
+            nextPoint = meshPoints[start + (pI + 1) % nPoints];
```

Comment:
Replaced the original line `nextPoint = meshPoints[f[fcIndex(pI)]];` with this line `nextPoint = meshPoints[start + (pI + 1) % nPoints];`. Is this Okay?